fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/NullableArrayTypeUtils.java

```diff
@@ -0,0 +1,119 @@
+import com.apple.foundationdb.record.query.plan.cascades.typing.Type;
+import com.apple.foundationdb.record.query.plan.cascades.values.MessageValue;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * A Utils class that holds logic related to nullable arrays.
+ */
+public class NullableArrayTypeUtils {
+    private static final String REPEATED_FIELD_NAME = "values";
+
+    private NullableArrayTypeUtils() {
+        throw new IllegalStateException("Utility class");
+    }
+
+    /**
+     * Get the fieldName of the repeated field.
+     *
+     * @return fieldName of the repeated field
+     */
+    public static String getRepeatedFieldName() {
+        return REPEATED_FIELD_NAME;
+    }
+
+    /**
+     * Check whether a descriptor describes a wrapped array. Message that looks like:
+     * message M {
+     * repeated R values = 1;
+     * }
+     * is considered to be a wrapped array.
+     *
+     * @param descriptor The protobuf descriptor.
+     *
+     * @return <code>true</code> if it describes a wrapped array, otherwise <code>false</code>.
+     */
+    public static boolean describesWrappedArray(@Nonnull Descriptors.Descriptor descriptor) {
+        if (descriptor.getFields().size() == 1) {
+            Descriptors.FieldDescriptor fieldDescriptor = descriptor.getFields().get(0);
+            return fieldDescriptor.isRepeated() && REPEATED_FIELD_NAME.equals(fieldDescriptor.getName());
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Check whether a nesting keyExpression is a wrapped array field.
+     *
+     * @param nestingKeyExpression the nesting keyExpression.
+     *
+     * @return <code>true</code> if it describes a wrapped array, otherwise <code>false</code>.
+     */
+    public static boolean isArrayWrapper(@Nonnull NestingKeyExpression nestingKeyExpression) {
+        KeyExpression child = nestingKeyExpression.getChild();
+        if (child.toKeyExpression().hasNesting()) {
+            RecordMetaDataProto.Field firstChild = child.toKeyExpression().getNesting().getParent();
+            return REPEATED_FIELD_NAME.equals(firstChild.getFieldName()) && RecordMetaDataProto.Field.FanType.FAN_OUT.equals(firstChild.getFanType());
+        }
+        return false;
+    }
+
+    /**
+     * Unwrap nested array in keyExpression.
+     *
+     * @param nestingKeyExpression The input keyExpression
+     *
+     * @return a keyExpression without wrapped array
+     */
+    @Nonnull
+    public static NestingKeyExpression unwrapArrayInKeyExpression(@Nonnull NestingKeyExpression nestingKeyExpression) {
+        final FieldKeyExpression parent = nestingKeyExpression.getParent();
+        final KeyExpression child = nestingKeyExpression.getChild();
+
+        RecordMetaDataProto.Nesting.Builder newNestingBuilder = RecordMetaDataProto.Nesting.newBuilder()
```

Comment:
Do you really need to build a new `KeyExpression`? Isn't the _new_ key expression you want to synthesize a part of the old one that is passed in?

Reply:
No, this is basically remove the "values" in the `KeyExpression`, so for structures like `a.values.b` we'll need to return `a.b`.

Reply:
What I mean is that there is no caller for this anymore is there? Whenever you are in need of doing this you are actually in the process of translating this into a query graph, so you should never actually have someone wanting a new key expression. TLDR: you should be able to remove this method altogether.